RFS/src/main/java/com/rfs/worker/IndexStep.java

```diff
@@ -0,0 +1,475 @@
+                 * 
+                 * If it's already exceeded the number of attempts, we attempt to mark it as failed.  If someone else
+                 * has updated the entry in the meantime, we just move on to the next work item.  This is safe because
+                 * it means someone else has either marked it as completed or failed, and either is fine.
+                 * 
+                 * If we fail to migrate it, we attempt to increment the attempt count.  It's fine if the increment
+                 * fails because we guarantee that we'll attempt the work at least N times, not exactly N times.
+                 */
+                if (workItem.numAttempts > CmsEntry.IndexWorkItem.ATTEMPTS_SOFT_LIMIT) {
+                    logger.info("Index Work Item " + workItem.name + " has exceeded the maximum number of attempts; marking it as failed...");
+                    CmsEntry.IndexWorkItem updatedEntry = new CmsEntry.IndexWorkItem(
+                        workItem.name,
+                        CmsEntry.IndexWorkItemStatus.FAILED,
+                        workItem.numAttempts,
+                        workItem.numShards
+                    );
+
+                    members.cmsClient.updateIndexWorkItem(updatedEntry, workItem).ifPresentOrElse(
+                        value -> logger.info("Index Work Item " + workItem.name + " marked as failed"),
+                        () ->logger.info("Unable to mark Index Work Item " + workItem.name + " as failed")
+                    );
+                    continue;
+                }
+
+                try {
+                    logger.info("Migrating index: " + workItem.name);
+                    IndexMetadata.Data indexMetadata = members.metadataFactory.fromRepo(members.snapshotName, workItem.name);
+
+                    ObjectNode root = indexMetadata.toObjectNode();
+                    ObjectNode transformedRoot = members.transformer.transformIndexMetadata(root);
+
+                    IndexMetadataData_OS_2_11 indexMetadataOS211 = new IndexMetadataData_OS_2_11(transformedRoot, indexMetadata.getId(), workItem.name);
+                    members.indexCreator.create(workItem.name, indexMetadataOS211).ifPresentOrElse(
+                        value -> logger.info("Index " + workItem.name + " created successfully"),
+                        () -> logger.info("Index " + workItem.name + " already existed; no work required")
+                    );
+
+                    logger.info("Forcefully updating the Index Work Item to indicate it has been completed...");
+                    CmsEntry.IndexWorkItem updatedEntry = new CmsEntry.IndexWorkItem(
+                        workItem.name,
+                        CmsEntry.IndexWorkItemStatus.COMPLETED,
+                        workItem.numAttempts,
+                        workItem.numShards
+                    );
+                    members.cmsClient.updateIndexWorkItemForceful(updatedEntry);
```

Comment:
I'm trying to wrap my head around treating index creation at this level of atomicity.  I understand that creating an index could be an expensive operation and could take some time.  However, that's going to be consuming resources mostly on the cluster, not the client.  The client would only be bound by the number of connections that it could make & that might be at 60K.  Is there a disadvantage to having one work item w/ a lease to do all of the index creation (& have the client do that creation in parallel)?

I know that we've gone back and forth on when to fanout work across nodes.  Snapshotting is definitely a single client task.  Templates and index migrations may need to happen in one order vs another and maybe only certain components will be pulled from a snapshot at a time depending on a user's desires.  Transformations may need to be tweaked to do them multiple times, possibly deleting ones that weren't created properly.  

Managing all of that for parallel distribution seems like it could be overkill.  Having phases to know when we are done and can move on is critical, but this is an area of code that could be a good spot to pickup external contributors.  To increase the likelihood of that, it will need to be as simple and exemplary as possible.

Reply:
Per discussion - we decided to simplify index creation by just having the worker that performs the setup do the index migrations too.